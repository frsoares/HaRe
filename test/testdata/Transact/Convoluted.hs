module Convoluted where

import Control.Concurrent


swapMVars :: MVar a -> MVar a -> IO ()
swapMVars x y = do
    a <- takeMVar x
    b <- takeMVar y
    putMVar x b
    putMVar y a




main :: IO ()
main = do
    v1 <- newEmptyMVar
    v2 <- newEmptyMVar
    v3 <- newEmptyMVar

    -- ...

    swapMVars v1 v2

    -- ...

    swapMVars v2 v3
