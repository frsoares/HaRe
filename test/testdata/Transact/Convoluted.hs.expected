module Convoluted where

import Control.Concurrent
import Control.Concurrent.STM


swapMVars :: TMVar a -> TMVar a -> IO ()
swapMVars x y = do
    a <- atomically (takeTMVar x)
    b <- atomically (takeTMVar y)
    atomically (putTMVar x b)
    atomically (putTMVar y a)




main :: IO ()
main = do
    v1 <- newEmptyTMVarIO
    v2 <- newEmptyTMVarIO
    v3 <- newEmptyTMVarIO

    -- ...

    swapMVars v1 v2

    -- ...

    swapMVars v2 v3
